keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -883,20 +889,43 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
 However, as disabling taint-manager is deprecated it is not a concern for this
 KEP.
 
-For Alpha, we implemented a fix for this issue by setting the pod phase as `Failed`
-in podgc.
-
-For Beta, we are going to simplify the code in job controller responsible for
-detecting pod failures. For this purpose we need to make sure that pods stuck
-in the pending phase which are terminating (with set deletionTimestamp) are
-eventually marked as failed. Such pods may either be not scheduled
-(for example when they are unschedulable due to node affinity configuration for
-which no node exists in the cluster) or scheduled (for example when they fail
-to pull the the docker image). In case of non-scheduled pods they will be moved
-to the `Failed` phase by the PodGC controller as it marks as failed all
-non-scheduled terminating pods (after the Alpha changes for this feature, see
-above). We plan to extend the PodGC controller to also move to the failed phase
-(and delete) scheduled pods in the pending state which are terminating.
+A scheduled pod may also get stuck in a `Pending` phase (for example due to an
+invalid image reference). In that case, as long as the pod is not Terminating it
```

Comment:
will kubelet retry image download for the pod with `restartPolicy=Never`? I simply don't remember off the top of my head

Reply:
will pod admission failure be in the same category of issues

Reply:
> will kubelet retry image download for the pod with `restartPolicy=Never`? I simply don't remember off the top of my head

Yes, Kubelet attempts to retry image download for as long as it either succeeds or the pod is terminating. I've tested this on KIND by blocking traffic with `iptables` and re-opening the traffic after 2min (after 4 failed attempts). Once re-opened the traffic, the image was pulled and the container started to run.

> will pod admission failure be in the same category of issues

No, when admission failure happens the pod is evicted (marked as `Failed` by Kubelet). More specifically, `rejectPod` is called: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2360-L2361, which sets the phase as `Failed`: https://github.com/kubernetes/kubernetes/blob/7b243cef1a81f4a0ac1619f7b9238fcd2bac2b9c/pkg/kubelet/kubelet.go#L2062-L2065